flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java

```diff
@@ -0,0 +1,382 @@
+        return DataTypes.STRUCTURED(
+                PercentileAccumulator.class,
+                DataTypes.FIELD(
+                        "percentages",
+                        DataTypes.ARRAY(DataTypes.DOUBLE()).bridgedTo(double[].class)),
+                DataTypes.FIELD("totalCount", DataTypes.BIGINT()),
+                DataTypes.FIELD(
+                        "valueCount",
+                        MapView.newMapViewDataType(DataTypes.DOUBLE(), DataTypes.BIGINT())));
+    }
+
+    @Override
+    public PercentileAccumulator createAccumulator() {
+        final PercentileAccumulator acc = new PercentileAccumulator();
+        acc.percentages = null;
+        acc.totalCount = 0;
+        acc.valueCount = new MapView<>();
+        return acc;
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // accumulate methods
+    // --------------------------------------------------------------------------------------------
+
+    public void accumulate(PercentileAccumulator acc, @Nullable Object value, Double percentage)
+            throws Exception {
+        if (acc.percentages == null) {
+            acc.setPercentages(percentage);
+        }
+        update(acc, value, 1L);
+    }
+
+    public void accumulate(
+            PercentileAccumulator acc,
+            @Nullable Object value,
+            Double percentage,
+            @Nullable Number frequency)
+            throws Exception {
+        if (acc.percentages == null) {
+            acc.setPercentages(percentage);
+        }
+        if (frequency == null || frequency.longValue() <= 0) {
+            return;
+        }
+        update(acc, value, frequency.longValue());
+    }
+
+    public void accumulate(PercentileAccumulator acc, @Nullable Object value, Double[] percentage)
+            throws Exception {
+        if (acc.percentages == null) {
+            acc.setPercentages(percentage);
+        }
+        update(acc, value, 1L);
+    }
+
+    public void accumulate(
+            PercentileAccumulator acc,
+            @Nullable Object value,
+            Double[] percentage,
+            @Nullable Number frequency)
+            throws Exception {
+        if (acc.percentages == null) {
+            acc.setPercentages(percentage);
+        }
+        if (frequency == null || frequency.longValue() <= 0) {
+            return;
+        }
+        update(acc, value, frequency.longValue());
+    }
+
+    private void update(PercentileAccumulator acc, @Nullable Object value, long frequency)
+            throws Exception {
+        if (value == null) {
+            return;
+        }
+
+        double val =
+                (value instanceof Number)
+                        ? ((Number) value).doubleValue()
+                        : DecimalDataUtils.doubleValue((DecimalData) value);
+
+        long cnt = Optional.ofNullable(acc.valueCount.get(val)).orElse(0L) + frequency;
+        if (cnt < 0) {
```

Comment:
We can't simpliy ignore this case, should update the count into state because there may exists non-sequential update inputs.

Reply:
Similar logic can be found in `MaxWithRetractAccumulator`